Docs model derives

notryanb opened the pull request:
Based on #1107

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,769 @@
+
+Diesel lets you write safe queries by using traits that you can implement on your models. Writing these implementations by hand can be very laborious, so in addition to the `diesel` crate, the `diesel_codegen` crate offers custom derives. In this section, we will cover each trait in detail in terms of its use cases, usage considerations, as well compare hand written implementations to its derive. ...
+
+Throughout this guide, we will be looking at examples for each trait and how they interact together. Some of the code will be implementing basic CRUD operations, however the details of those operations will be not be covered beyond their relevance to the demonstrated trait.
```
explain/link to explanation of CRUD?

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,769 @@
+
+## <a name="queryable"></a>Queryable
+
+Using the [#[derive(Queryable)]](http://docs.diesel.rs/diesel/query_source/trait.Queryable.html) trait annotation on your model struct allows records to be queried from the database. This means that any time you are using the `load()`, `get_result()`, `get_results()`, and `first()` methods to execute your queries, you must have `Queryable` implemented on your model.
```
I'd put the derive(…) in back ticks, and make the link a ref link so humans can better parse this :)

killercup replied:
That last sentence might be confusing, because the Queryable trait does not add these methods. But I'm not sure how to better express this without an aside on how ExecuteDsl and LoadDsl are implemented…

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,769 @@
+}
+```
+
+When deriving Queryable for a model, the struct must have all fields defined. Queryable's associated type, Row,  implements `FromSqlRow`, which provides the `fields_needed()` method. That method returns the number of fields to be consumed by the row. If we were to comment out all the three `String` fields on our User struct, we would see the following error.
````
back ticks around Queryable, Row, maybe also User

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,769 @@
+fn main() {
+    // The following will return all Users as a QueryResult<Vec<User>>
+    let query_result = users.load::<User>(&db_connection)
+      .expect("Error loading users");
```
2 spaces indent missing?

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,769 @@
+// externs, database connection, and using statement code omitted
+
+fn main() {
+    // The following will return all Users as a QueryResult<Vec<User>>
```
That comment is good, but you can also just annotate the let binding and give it a speaking name:

```rust
let users_result: QueryResult<Vec<User>> = user.load(&db_connection);
```

This even gets rid of the turbofish (`::<User>`)

If you want t talk about the QueryResult type, use that, and maybe even add

```rust
let users = users_result.expect("Error loading users");
```

Otherwise, change it to `let users: Vec<User>` and just append the `expect`.

notryanb replied:
I definitely like talking about the `QueryResult` and I think this point will be illustrated better by type annotating the let binding like you have shown. 

My only concern is for the current guide on the Diesel website. All the examples I've seen use the turbofish and I would just like to maintain consistency. Is one of the ways more idiomatic || rusty?

killercup replied:
[Historically](https://www.reddit.com/r/rust/comments/3fimgp/why_double_colon_rather_that_dot/ctozkd0/), people seem to like the name 'turbofish' more than its look. I personally prefer variable type annotations more than putting a turbofish at the end of a long chain of method calls, because you put the name right next to the type. But others may disagree, as the turbofish version of something like `.load::<Post>(&c)` or `.parse::<i32>()` can be read quite nicely as "load posts from connections" or "parse this as an integer".

You know what? I'll open an issue on the [fmt-rfcs](https://github.com/rust-lang-nursery/fmt-rfcs) repo.

notryanb replied:
Awesome! 😄

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,769 @@
+  --> src/bin/main.rs:21:28
+   |
+21 |     let query_result = users.load::<User>(&db_connection).expect("Error loading users");
+   |                            ^^^^ the trait `diesel::types::FromSqlRow<(diesel::types::Integer, diesel::types::Text, diesel::types::Text, diesel::types::Text), _>` is not implemented for `(i32,)`
```
Where should the `^^^^` be? Underneath `load`? Seems a bit off, at least in Github's diff view

notryanb replied:
I'll look into the formatting for this. It is under `.load`. Looks fine in vim && my markdown preview, but there might be some inconsistent spacing issues from when I pasted it from the cli.

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,769 @@
+   = note: required because of the requirements on the impl of `diesel::LoadQuery<_, diesel_demo_cli::models::User>` for `diesel_demo_cli::schema::users::table`
+```
+
+Notice the compiler is indicating a trait is not implmented for using the `.load()` method. FromSqlRow is expecting a tuple that looks like `(i32, std::string::String, std::string::String, std::string::String)`, but we currently only have a tuple consisting of `(i32,)`.
````
Do you want to talk a bit more about how to parse that error message? I don't think it's necessary, but I talked a bit about that [here](https://deterministic.space/hook-into-rustc-errors.html), feel free to adapt/copy it.

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,769 @@
+    }
+}
+
+
```
✂️ empty lines here and at the start of the code block

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,769 @@
+
+```
+
+The advantages of using the derive annotation should be clear here, as the more columns you have, the more code you would have to write. You can thank Rust for this zero-cost-abstraction!
````
"zero-cost abstraction" with only one dash.

Also: Is code generation/macros a zero-cost abstraction? https://blog.rust-lang.org/2015/05/11/traits.html quotes

> What you don’t use, you don’t pay for [Stroustrup, 1994]. And further: What you do use, you couldn’t hand code any better.

Seems ~~abstract~~ ambiguous to me in this case

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,769 @@
+
+## <a name="insertable"></a>Insertable
+
+Using the [#[derive(Insertable)]](http://docs.diesel.rs/diesel/prelude/trait.Insertable.html) trait annotation on your model struct allows records to be inserted into the database via that struct. You will need `Insertable` when using the `diesel::insert(&my_new_row).into(my_table::table)` API. You may pass a reference to a single struct, `Vec` or slice to insert a single record or multiple recor ...
```
Same as above, derive in back ticks, ref link style link.

killercup replied:
Also, maybe "Adding… to…" instead of "Using… on…" here?

killercup replied:
How about:

> You may pass a reference to a single struct to insert a single record. Pass a `Vec` or a slice to insert multiple records at once.

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,769 @@
+
+Using the [#[derive(Insertable)]](http://docs.diesel.rs/diesel/prelude/trait.Insertable.html) trait annotation on your model struct allows records to be inserted into the database via that struct. You will need `Insertable` when using the `diesel::insert(&my_new_row).into(my_table::table)` API. You may pass a reference to a single struct, `Vec` or slice to insert a single record or multiple recor ...
+
+When implementing `Insertable`, you probably won't be setting the auto-incremented `Id` field of the row. For this reason, it's not advisable to use `Queryable` and `Insertable` on the same struct due to the field number constraints of `Queryable`. Create another struct that you may use for database insertions that will have all the fields you would like to set. This section will not cover nullab ...
```
`Id` field name is typically lowercase

killercup replied:
Other than `id`, timestamp fields like `created_at` and `updated_at` are also usually omitted in the Insertable structs

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,769 @@
+
+#[derive(Insertable)]
+#[table_name="users"]
+pub struct NewUser<'a> {
```
Nice touch with the lifetime!

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,769 @@
+   |
+31 |     diesel::insert(&new_user).into(users::table)
+   |                              ^^^^ the trait `diesel::query_builder::insert_statement::IntoInsertStatement<_, diesel::query_builder::insert_statement::Insert>` is not implemented for `&diesel_demo_cli::models::NewUser<'_>`
+```
````
Add empty line below code block

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,769 @@
+31 |     diesel::insert(&new_user).into(users::table)
+   |                              ^^^^ the trait `diesel::query_builder::insert_statement::IntoInsertStatement<_, diesel::query_builder::insert_statement::Insert>` is not implemented for `&diesel_demo_cli::models::NewUser<'_>`
+```
+In the above error, the compiler is telling us there is an unsatisfied trait bound from IntoInsertStatement. `diesel::insert()` returns an `IncompleteInsertStatement`, which implements `into()`. `into()` specifies the table the `Insertable` data should be passed to, which is why this constraint is reported to us.
````
> which is why this constraint is reported to us.

Can you make this a bit more… approachable? An extreme way might be:

> So, when you pass it something that is not `Insertable`, Rust gets confused and asks for your help to make it work.

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,769 @@
+```
+In the above error, the compiler is telling us there is an unsatisfied trait bound from IntoInsertStatement. `diesel::insert()` returns an `IncompleteInsertStatement`, which implements `into()`. `into()` specifies the table the `Insertable` data should be passed to, which is why this constraint is reported to us.
+
+The example below is how we might implement `Insertable` without the derive annotation. In this case, we are using the `impl_Insertable!` macro, but there are a few more ways which enable  us from having to reimplement our `NewUser` struct.
````
> enable  us

double space

Also… this whole sentence is confusing. And using the macro is cheating! 😝

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,769 @@
+}
+
+
+```
````
✂️  empty lines at start/end of code block

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,769 @@
+```
+
+
+## <a name="identifiable"></a>Identifiable
````
(Will continue to review from here later)

notryanb commented:
@killercup Thanks for all the input! I'll start working on these changes.

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+
+## <a name="queryable"></a>Queryable
+
+Using the [`#[derive(Queryable)]`](http://docs.diesel.rs/diesel/query_source/trait.Queryable.html)` trait annotation on your model struct allows records to be queried from the database.
```
stray back tick

Also, i'd use ref links everywhere, i.e.

```markdown
Using the [`#[derive(Queryable)]`] trait annotation on your model struct allows records to be queried from the database.

[`#[derive(Queryable)]`]: http://docs.diesel.rs/diesel/query_source/trait.Queryable.html
```

which make this easier to read and catch syntax stuff like that

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+
+Using the [`#[derive(Queryable)]`](http://docs.diesel.rs/diesel/query_source/trait.Queryable.html)` trait annotation on your model struct allows records to be queried from the database.
+This means that any time you are using the `load()`, `get_result()`, `get_results()`, and `first()` methods to execute your queries, you must have `Queryable` implemented on your model. 
+`Queryable` doesn't directly provide these methods, but they are a "side effect" of the implementation.
```
Hm, side effect is quite loaded, even in quotes. Maybe

> `Queryable` doesn't directly provide these methods, but the traits these methods come from (and that part of Diesel's prelude) require your type to implement `Queryable`.

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+You must include all the table's fields in your model struct.
+The implementation details for this rule are only important if you wish to `impl` the trait by hand and will be covered later in this section.
+Your struct name should be the singular of its corresponding table name.
+Queryable by itself does not support the `#[table_name=""]` attribute. 
```
Maybe combine these four sentences with conjunctions? They read like separate yet similar lists of facts.

notryanb replied:
I originally thought this should be a bulleted list for a quick reference, but that is not how the rest of the guide is structured. 

How about...?
>  There are some considerations to be made when implementing the `Queryable` trait. First, you must include all the table's fields in your model struct. `Queryable` needs to know how many columns are in each row in order to `build()` that row. Also, your struct should be named in the singular form of its corresponding table name, as `Queryable` does not support the `#[table_name=""]` attribute.

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+    // The following will return all Users as a QueryResult<Vec<User>>
+    let users_result: QueryResult<Vec<User>> = users.load(&db_connection);
+
+    // Here we are getting the value(or error) out of the QueryResult
```
```
value (or error)
     ^
```

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,926 @@
+```
+
+Notice the compiler is indicating a trait is not implmented for using the `.load()` method.
+When reading, take note of the values in the tuple.
````
"tuple[s]"

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+
+Notice the compiler is indicating a trait is not implmented for using the `.load()` method.
+When reading, take note of the values in the tuple.
+`FromSqlRow<(Integer, Text, Text, Text, Text)>` not implemented for `(i32,`).
```
Maybe write that quote from the error message as block quote?

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+`FromSqlRow` is trying to convert those 4 types into the types on your model struct.
+Our model struct has the three String fields commented out, so it doesn't know what those `Text` columns are supposed to be converted to.
+In other words,
+FromSqlRow is expecting a tuple that looks like `(i32, std::string::String, std::string::String, std::string::String)`,
```
You can omit the `std::string::`s here

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,926 @@
+   |                               ^^^^ the trait `diesel::query_builder::insert_statement::IntoInsertStatement<_, diesel::query_builder::insert_statement::Insert>` is not implemented for `&diesel_demo_cli::models::NewUser<'_>`
+```
+
+In the above error, the compiler is telling us there is an unsatisfied trait bound from IntoInsertStatement.
````
IntoInsertStatement in back ticks

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+When using the `Indentifiable` on a struct,
+it means that struct can be identified on a single table in the database. 
+
+Your struct *must* have an `id` field on it.
```
Doesn't that conflict with

> If your table's primary key is named differently, you can annotate the table with the attribute

notryanb replied:
whoops! That will be omitted. The next line is what I was trying to convey.

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+      first_name: "Gordon", 
+      last_name: "Freeman", 
+      email: "gordon.freeman@blackmesa.co" 
+  };
```
[now we can actually merge it](https://gitter.im/diesel-rs/diesel?at=5990b9e51c8697534ab4b118)

notryanb replied:
I want to make sure this PR abides by Diesel (unofficial) Contributing docs. :smile:

killercup commented on guide_drafts/model-derives.md:
````diff
@@ -0,0 +1,926 @@
+}
+```
+
+## <a name="aschangeset"></a>AsChangeset
````
memo to self: read up to this section

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+## <a name="aschangeset"></a>AsChangeset
+
+For more slightly more complicated database updates,
+`AsChangeset` can make your development easier.
```
> make your development easier

it that something people say? Would've gone with "make your code more ergonomic"

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+`AsChangeset` can make your development easier.
+In this section we will change our schema to make our `email` field nullable.
+The nullable fields on our model structs will now be of type `Option<T>`.
+You will probably be not changing the primary key of the row or rows that you're updating.
```
how about "Usually, you don't want to change…"

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+In this section we will change our schema to make our `email` field nullable.
+The nullable fields on our model structs will now be of type `Option<T>`.
+You will probably be not changing the primary key of the row or rows that you're updating.
+For this reason, we probably won't have `AsChangeset` and `Queryable` annotated on the same struct,
```
"probably won't have" -> "don't want to have … here,"

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,926 @@
+The nullable fields on our model structs will now be of type `Option<T>`.
+You will probably be not changing the primary key of the row or rows that you're updating.
+For this reason, we probably won't have `AsChangeset` and `Queryable` annotated on the same struct,
+which means we'll need another struct for updating our records.
```
Maybe even add this somewhere? I don't think we ever wrote that down somewhere prominently:

> It's considered good practice in diesel to have one struct per "usage". E.g., one struct to query users, one to create new users, one to change just the email, etc. This often corresponds to the way applications access the database. Don't be afraid to have multiple structs per database table—they are not like the classes that correspond to tables 1:1 in other ORMs.

notryanb replied:
I think this is a really important point. I'm coming from a rails background, so this seemed alien at first. I'll think about a good spot to add it.

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,935 @@
+Both the parent and child struct must have the [Identifiable](#identifiable) trait implemented.
+By default, the foreign key column on the child should take the form of `parent_id`.
+If there is a custom foreign key,
+the belongs_to attribute be written as `#[belongs_to(ParentStruct, foreign_key="my_custom_key")]`.
```
belongs_to in back ticks

Oh, and add a verb somewhere, e.g. "must" ;)

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,935 @@
+           `&mut diesel_demo::models::Post : diesel::associations::BelongsTo<diesel_demo::models::User>`
+           `&mut diesel_demo::models::Post : diesel::associations::HasTable`
+           `&mut diesel_demo::models::Post : diesel::associations::BelongsTo<diesel_demo::models::User>`
+           `&mut diesel_demo::models::Post : diesel::associations::BelongsTo<diesel_demo::models::User>`
```
Yeah you can probably replace the last 37 lines with ellipses

notryanb replied:
😁

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,935 @@
+
+In this particular case,
+we are using the `belongs_to()` method, which requires us to implement the `BelongsTo` trait.
+Below is what `cargo expand` generates for `#[derive(Associations)]`.
```
Btw, is there a note about this mysterious `cargo expand` anywhere? I'm not sure many people know of it

killercup commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,935 @@
+    pub first_name: &'a str,
+    pub last_name: &'a str,
+    pub email: Option<Option<&'a str>>,
+ }
```
I you use the "let's just update the name because email requires double opt-in" use case it makes more sense because then the struct is _not_ the same as `NewUser` :)

killercup replied:
Ah, forget it, you can't show the `Option<Option<_>>` then. Sorry, commented too fast.

notryanb commented:
Revisions from this weekend
- Implemented ref links
- Added a small conclusion section
- Changed some verbiage in a few sections, removed some redundant points.
- Fixed a few markdown formatting issues.

sgrif commented on guide_drafts/model-derives.md:
```diff
@@ -0,0 +1,930 @@
+
+When deriving `Queryable` for a model, the struct must have all fields defined.
+`Queryable`'s associated type, `Row`,  implements `FromSqlRow`, which provides the `fields_needed()` method.
+That method returns the number of fields to be consumed by the row.
```
This doesn't seem relevant to users of `Queryable`

notryanb commented:
@killercup @sgrif - I made some small changes to `Queryable`. I felt I wasn't conveying it's purpose correctly, so I changed some verbiage and updated the example.

killercup commented:
Are you all afraid of this small green button? I say :shipit:
